http/src/main/scala/org/http4s/blaze/http/http2/server/ServerStage.scala

```diff
@@ -0,0 +1,144 @@
+import org.http4s.blaze.util.{BufferTools, Execution}
+
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.Future
+import scala.util.{Failure, Success, Try}
+
+/** Basic implementation of a http2 stream [[TailStage]] */
+private[http] class ServerStage(
+  streamId: Int,
+  service: HttpService,
+  config: HttpServerStageConfig
+) extends TailStage[StreamMessage] {
+
+  private implicit def _ec = Execution.trampoline // for all the onComplete calls
+
+  private val timeoutService = ServiceTimeoutFilter(config.serviceTimeout)(service)
+
+  override def name = s"Http2StreamStage($streamId)"
+
+  override protected def stageStartup(): Unit = {
+    super.stageStartup()
+    startRequest()
+  }
+
+  private def shutdownWithCommand(cmd: Cmd.OutboundCommand): Unit = {
+    stageShutdown()
+    sendOutboundCommand(cmd)
+  }
+
+  private def startRequest(): Unit = {
+    // The prelude should already be available (or we wouldn't have a stream id)
+    // so adding a timeout is unnecessary.
+    channelRead().onComplete  {
+      case Success(HeadersFrame(_, endStream, hs)) =>
+        if (endStream) checkAndRunRequest(hs, BodyReader.EmptyBodyReader)
+        else getBodyReader(hs)
+
+      case Success(frame) =>
+        val e = PROTOCOL_ERROR.goaway(
+          s"Stream $streamId received invalid frame: ${frame.getClass.getSimpleName}")
+        shutdownWithCommand(Cmd.Error(e))
+
+        // TODO: what about a 408 response for a timeout?
+      case Failure(Cmd.EOF) => shutdownWithCommand(Cmd.Disconnect)
+
+      case Failure(t) =>
+        logger.error(t)("Unknown error in startRequest")
+        val e = INTERNAL_ERROR.rst(streamId, s"Unknown error")
+        shutdownWithCommand(Cmd.Error(e))
+    }(Execution.directec)
+  }
+
+  private[this] def getBodyReader(hs: Headers): Unit = {
+    val length: Option[Try[Long]] = hs.collectFirst {
+      case (ContentLength, v) =>
+        try Success(java.lang.Long.valueOf(v))
+        catch { case t: NumberFormatException =>
+           Failure(PROTOCOL_ERROR.rst(streamId, s"Invalid content-length: $v."))
+        }
+    }
+
+    length match {
+      case Some(Success(len)) => checkAndRunRequest(hs, new BodyReaderImpl(len))
+      case Some(Failure(error)) => shutdownWithCommand(Cmd.Error(error))
+      case None => checkAndRunRequest(hs, new BodyReaderImpl(-1))
+    }
+  }
+
+  private[this] def checkAndRunRequest(hs: Headers, bodyReader: BodyReader): Unit = {
+    RequestParser.makeRequest(hs, bodyReader) match {
+      case Right(request) =>
+        timeoutService(request).onComplete(renderResponse(request.method, _))(config.serviceExecutor)
+
+      case Left(errMsg) =>
+        shutdownWithCommand(Cmd.Error(PROTOCOL_ERROR.rst(streamId, errMsg)))
+    }
+  }
+
+  private[this] def renderResponse(method: String, response: Try[RouteAction]): Unit = response match {
+    case Success(builder) =>
+      builder.handle(getWriter(method == "HEAD", _))
```

Comment:
Its clear in the context of the tests, but not clear at the call that this == is the check that is forcing the NoOp writer on a head request. Once you go to the other functions its more understandable, but it would nice to be clearer at the invocation site.

Reply:
Thats a fair point, and I vaguely remember thinking it was a bit tricky at some point as well. I'll make the code a hair more verbose and/or add some comments to clear it up.